iree/compiler/Dialect/Stream/Transforms/MaterializeCopyOnWrite.cpp

```diff
@@ -0,0 +1,186 @@
+  // Any ops we materialize must have the same affinity as their consumer. This
+  // ensures the copies we issue happen locally to the consumer.
+  IREE::Stream::AffinityAttr affinity;
+  if (auto affinityOp =
+          dyn_cast<IREE::Stream::AffinityOpInterface>(tiedOp.getOperation())) {
+    affinity = affinityOp.getAffinity();
+  }
+
+  // Clones each operand that is tied to a result and it may be required.
+  OpBuilder builder(tiedOp);
+  unsigned tiedOperandsOffset = tiedOp.getTiedOperandsIndexAndLength().first;
+  auto tiedOperandIndices = tiedOp.getTiedResultOperandIndices();
+  for (unsigned i = 0; i < tiedOperandIndices.size(); ++i) {
+    int64_t operandIdx = tiedOperandIndices[i];
+    if (operandIdx == IREE::Util::TiedOpInterface::kUntiedIndex) continue;
+    auto &operand = tiedOp->getOpOperand(tiedOperandsOffset + operandIdx);
+    didChange =
+        materializeOperandCOW(tiedOp.getLoc(), operand, affinity, builder) ||
+        didChange;
+  }
+
+  return didChange;
+}
+
+// Materializes copies on writes within |region|.
+// Returns true if any copy was required and materialized.
+static bool materializeRegionCOW(Region &region) {
+  bool didChange = false;
+  for (auto &block : region.getBlocks()) {
+    for (auto &op : block) {
+      if (!op.hasTrait<OpTrait::IREE::Stream::AsyncPhaseOp>()) continue;
+      didChange =
+          TypeSwitch<Operation *, bool>(&op)
+              .Case<IREE::Stream::TensorImportOp, IREE::Stream::TensorExportOp,
+                    IREE::Stream::AsyncFillOp, IREE::Stream::AsyncUpdateOp,
+                    IREE::Stream::AsyncCopyOp, IREE::Stream::AsyncDispatchOp,
+                    IREE::Stream::AsyncExecuteOp,
+                    IREE::Stream::AsyncConcurrentOp>(
+                  [&](auto op) { return materializeTiedOpCOW(op); })
+              .Default(false) ||
+          didChange;
+    }
+  }
+  return didChange;
+}
+
+//===----------------------------------------------------------------------===//
+// -iree-stream-materialize-copy-on-write
+//===----------------------------------------------------------------------===//
+
+// Applies a relatively simple heuristic to insert copies where they _may_ be
+// required. This may introduce copies that are not required for the sake of
+// ensuring correctness. Intended to be paired with
+// -iree-stream-elide-async-copies.
+//
+// Conceptually this work is performed in two phases: copy insertion and copy
+// elision. This pass inserts copies at all mutation sites regardless of whether
+// they are required, effectively disabling ties as a mechanism for in-place
+// updates but ensuring correct execution semantics. Afterward a dataflow
+// analysis pass is run to identify which copies can be elided based on use-def
+// chains (including ones spanning the CFG). Though this process can lead to
+// additional copies it is easier to ensure that each pass works independently
+// and also makes it easy to disable copy elision to ferret out issues.
```

Comment:
This ("also makes it easy to disable copy elision to ferret out issues") is a little at odds with

>   // pass attempting to do both. Note that copy-on-write materialization is
  // required for correct execution while copy elision is for performance only
  // (though it's critical enough that it is not optional).

Having an option to disable the pass would make it easy to disable (no source changes needed).

Reply:
yeah we can add pipeline options for these things (like scheduling concurrent execution as well) - of course, setting the pipeline options themselves is tricky plumbing that I still don't think has any good solution besides a bunch of boilerplate